File: screener/criteria/plugins/cheap_quality.py
"""Value + Quality: P/E <20, ROE >15%, low debt, bullish trend."""

from __future__ import annotations

from tradingview_screener import col

from screener.criteria import criterion


@criterion("cheap_quality")
def cheap_quality() -> list:
    return [
        col("price_earnings_ttm") > 0,
        col("price_earnings_ttm") <= 20,
        col("return_on_equity") > 15,
        col("debt_to_equity") < 1,
        col("EMA20") > col("EMA200"),
    ]

File: screener-rs/src/screeners/criteria/cheap_quality.rs
use super::{CmpOp, Predicate, cmp, field, val};

pub const NAME: &str = "cheap_quality";

pub fn predicates() -> Vec<Predicate> {
    vec![
        cmp(field("price_earnings_ttm"), CmpOp::Gt, val(0.0)),
        cmp(field("price_earnings_ttm"), CmpOp::Lte, val(20.0)),
        cmp(field("return_on_equity"), CmpOp::Gt, val(15.0)),
        cmp(field("debt_to_equity"), CmpOp::Lt, val(1.0)),
        cmp(field("EMA20"), CmpOp::Gt, field("EMA200")),
    ]
}
